c8aee0d8 (#2911) makes create_mirror_view_and_copy() trigger a static_assert about copying to const

Before revision c8aee0d8, if view v has a const value type, it was allowed to do:
```
auto v_host = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), v);
```
But after, this code causes a static_assert to fail:
```
/ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/unit_test/graph/Test_Graph_graph_color_distance2.hpp:420:1:   required from here
/home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_CopyViews.hpp:1756:72: error: static assertion failed: deep_copy requires non-const destination type
 1756 |                              typename dst_type::non_const_value_type>::value,
      |                                                                        ^~~~~
/home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_CopyViews.hpp:1932:21: error: invalid conversion from ‘const void*’ to ‘void*’ [-fpermissive]
 1932 |       Kokkos::Impl::DeepCopy<dst_memory_space, src_memory_space>(
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     |
      |                     const void*
 1933 |           dst.data(), src.data(), nbytes);
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

@dalg24 I copied the error message from the last failing step of git bisect, which was c8aee.
I think this commit was somewhere in the middle of PR #2911, so Kokkos_CopyViews.hpp was changed in some later commits. Here's an error message with develop (26b0118), saying line 1911 which matches the file you linked:

```
[ 65%] Building CXX object unit_test/CMakeFiles/KokkosKernels_blas_openmp.dir/openmp/Test_OpenMP_Blas1_dot.cpp.o
In file included from /home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_Core.hpp:96,
                 from /ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/unit_test/serial/Test_Serial.hpp:2,
                 from /ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/unit_test/serial/Test_Serial_Graph_graph_color_d2.cpp:1:
/home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_CopyViews.hpp: In instantiation of ‘void Kokkos::deep_copy(const Kokkos::View<D, P ...>&, const Kokkos::View<RT, RP ...>&, typename std::enable_if<((std::is_same<typename Kokkos::ViewTraits<DataType, Properties ...>::specialize, void>::value && std::is_same<typename Kokkos::ViewTraits<RT, RP ...>::specialize, void>::value) && (((unsigned int)(Kokkos::ViewTraits<DataType, Properties ...>::rank) != 0) || ((unsigned int)(Kokkos::ViewTraits<RT, RP ...>::rank) != 0)))>::type*) [with DT = const int*; DP = {Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>}; ST = const int*; SP = {Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>}; typename std::enable_if<((std::is_same<typename Kokkos::ViewTraits<DataType, Properties ...>::specialize, void>::value && std::is_same<typename Kokkos::ViewTraits<RT, RP ...>::specialize, void>::value) && (((unsigned int)(Kokkos::ViewTraits<DataType, Properties ...>::rank) != 0) || ((unsigned int)(Kokkos::ViewTraits<RT, RP ...>::rank) != 0)))>::type = void]’:
/home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_CopyViews.hpp:3564:12:   required from ‘typename Kokkos::Impl::MirrorViewType<Space, T, P>::view_type Kokkos::create_mirror_view_and_copy(const Space&, const Kokkos::View<D, P ...>&, const string&, typename std::enable_if<Kokkos::Impl::MirrorViewType<Space, T, P ...>::is_same_memspace>::type*) [with Space = Kokkos::HostSpace; T = const int*; P = {Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1>}; typename Kokkos::Impl::MirrorViewType<Space, T, P>::view_type = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >; std::string = std::__cxx11::basic_string<char>; typename std::enable_if<Kokkos::Impl::MirrorViewType<Space, T, P ...>::is_same_memspace>::type = void]’
/ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/src/graph/impl/KokkosGraph_Distance2Color_impl.hpp:873:102:   required from ‘void KokkosGraph::Impl::GraphColorDistance2<HandleType, rowmap_t, entries_t, doing_bipartite>::compute_d2_coloring_serial(const color_view_type&) [with HandleType = KokkosGraph::GraphColorDistance2Handle<const int, const int, const int, Kokkos::Serial, Kokkos::HostSpace, Kokkos::HostSpace>; rowmap_t = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >; entries_t = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >; bool doing_bipartite = false; KokkosGraph::Impl::GraphColorDistance2<HandleType, rowmap_t, entries_t, doing_bipartite>::color_view_type = Kokkos::View<int*, Kokkos::HostSpace>]’
/ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/src/graph/impl/KokkosGraph_Distance2Color_impl.hpp:207:13:   required from ‘void KokkosGraph::Impl::GraphColorDistance2<HandleType, rowmap_t, entries_t, doing_bipartite>::compute_distance2_color() [with HandleType = KokkosGraph::GraphColorDistance2Handle<const int, const int, const int, Kokkos::Serial, Kokkos::HostSpace, Kokkos::HostSpace>; rowmap_t = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >; entries_t = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >; bool doing_bipartite = false]’
/ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/src/graph/KokkosGraph_Distance2Color.hpp:96:3:   required from ‘void KokkosGraph::Experimental::graph_color_distance2(KernelHandle*, typename KernelHandle::nnz_lno_t, InRowmap, InEntries) [with KernelHandle = KokkosKernels::Experimental::KokkosKernelsHandle<int, int, double, Kokkos::Serial, Kokkos::HostSpace, Kokkos::HostSpace>; InRowmap = Kokkos::View<const int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; InEntries = Kokkos::View<int*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >; typename KernelHandle::nnz_lno_t = int]’
/ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/unit_test/graph/Test_Graph_graph_color_distance2.hpp:191:9:   required from ‘void test_dist2_coloring(lno_t, size_type, lno_t, lno_t) [with scalar_unused = double; lno_t = int; size_type = int; device = Kokkos::Serial]’
/ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/unit_test/graph/Test_Graph_graph_color_distance2.hpp:410:1:   required from here
/home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_CopyViews.hpp:1735:72: error: static assertion failed: deep_copy requires non-const destination type
 1735 |                              typename dst_type::non_const_value_type>::value,
      |                                                                        ^~~~~
/home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_CopyViews.hpp:1911:21: error: invalid conversion from ‘const void*’ to ‘void*’ [-fpermissive]
 1911 |       Kokkos::Impl::DeepCopy<dst_memory_space, src_memory_space>(
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                     |
      |                     const void*
 1912 |           dst.data(), src.data(), nbytes);
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_View.hpp:54,
                 from /home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_Parallel.hpp:53,
                 from /home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_Serial.hpp:56,
                 from /home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_Core.hpp:54,
                 from /ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/unit_test/serial/Test_Serial.hpp:2,
                 from /ascldap/users/bmkelle/KK_FixSpgemmColor/kokkos-kernels/unit_test/serial/Test_Serial_Graph_graph_color_d2.cpp:1:
/home/bmkelle/KK_FixSpgemmColor/ompBuild/kokkos-install/include/Kokkos_HostSpace.hpp:308:18: note:   initializing argument 1 of ‘Kokkos::Impl::DeepCopy<Kokkos::HostSpace, Kokkos::HostSpace, ExecutionSpace>::DeepCopy(void*, const void*, size_t) [with ExecutionSpace = Kokkos::OpenMP; size_t = long unsigned int]’
  308 |   DeepCopy(void* dst, const void* src, size_t n) {
      |            ~~~~~~^~~
```

--- a/core/src/Kokkos_CopyViews.hpp
+++ b/core/src/Kokkos_CopyViews.hpp
@@ -3560,8 +3560,7 @@ create_mirror_view_and_copy(
         Impl::MirrorViewType<Space, T, P...>::is_same_memspace>::type* =
         nullptr) {
   (void)name;
-  // Inherit fencing behavior of deep_copy
-  deep_copy(src, src);
+  fence();  // same behavior as deep_copy(src, src)
   return src;
 }
 
@@ -3576,7 +3575,8 @@ create_mirror_view_and_copy(
         !Impl::MirrorViewType<Space, T, P...>::is_same_memspace>::type* = 0) {
   using Mirror      = typename Impl::MirrorViewType<Space, T, P...>::view_type;
   std::string label = name.empty() ? src.label() : name;
-  auto mirror = Mirror(ViewAllocateWithoutInitializing(label), src.layout());
+  auto mirror       = typename Mirror::non_const_type{
+      ViewAllocateWithoutInitializing(label), src.layout()};
   deep_copy(mirror, src);
   return mirror;
 }

--- a/core/unit_test/TestViewAPI.hpp
+++ b/core/unit_test/TestViewAPI.hpp
@@ -841,6 +841,24 @@ struct TestViewMirror {
     ASSERT_EQ(a_org(5), a_h3(5));
   }
 
+  template <typename View>
+  static typename View::const_type view_const_cast(View const &v) {
+    return v;
+  }
+
+  static void test_mirror_copy_const_data_type() {
+    using ExecutionSpace = typename DeviceType::execution_space;
+    int const N          = 100;
+    Kokkos::View<int *, ExecutionSpace> v("v", N);
+    Kokkos::deep_copy(v, 255);
+    auto v_m1 = Kokkos::create_mirror_view_and_copy(
+        Kokkos::DefaultHostExecutionSpace(), view_const_cast(v));
+    ASSERT_TRUE(run_check(v_m1, 255));
+    auto v_m2 = Kokkos::create_mirror_view_and_copy(ExecutionSpace(),
+                                                    view_const_cast(v));
+    ASSERT_TRUE(run_check(v_m2, 255));
+  }
+
   template <class MemoryTraits, class Space>
   struct CopyUnInit {
     typedef typename Kokkos::Impl::MirrorViewType<
@@ -896,6 +914,7 @@ struct TestViewMirror {
     test_mirror_view<Kokkos::MemoryTraits<Kokkos::Unmanaged> >();
     test_mirror_copy<Kokkos::MemoryTraits<0> >();
     test_mirror_copy<Kokkos::MemoryTraits<Kokkos::Unmanaged> >();
+    test_mirror_copy_const_data_type();
     test_mirror_no_initialize<Kokkos::MemoryTraits<0> >();
     test_mirror_no_initialize<Kokkos::MemoryTraits<Kokkos::Unmanaged> >();
   }

--- a/core/unit_test/TestViewAPI.hpp
+++ b/core/unit_test/TestViewAPI.hpp
@@ -853,10 +853,8 @@ struct TestViewMirror {
     Kokkos::deep_copy(v, 255);
     auto v_m1 = Kokkos::create_mirror_view_and_copy(
         Kokkos::DefaultHostExecutionSpace(), view_const_cast(v));
-    ASSERT_TRUE(run_check(v_m1, 255));
     auto v_m2 = Kokkos::create_mirror_view_and_copy(ExecutionSpace(),
                                                     view_const_cast(v));
-    ASSERT_TRUE(run_check(v_m2, 255));
   }
 
   template <class MemoryTraits, class Space>

